### PR TITLE
F transferbetweenaccounts

### PR DESCRIPTION
**Pull request by xxshadowfang**
This Adds the functionality of Transfers between accounts. 

To add a transfer you add it via the add Transaction popup. 
Any edits in the table should be reflected in both transactions. 

Throw everyrhing at this code as there are a lot of edge scenarios.

### CODE REVIEW

File: src/main/java/ledger/database/entity/Type.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ public boolean equals(Object o) {
 
         Type type2 = (Type) o;
 
-        if (!(this.id == type2.getId())) return false;
+        //if (!(this.id == type2.getId())) return false;
```
Comment (shellajt) [inline src/main/java/ledger/database/entity/Type.java]: Lets get rid of this

File: src/main/java/ledger/database/storage/SQL/ISQLDatabaseTransaction.java [hunk 2]
```diff
@@ -84,7 +109,13 @@ default void deleteTransaction(Transaction transaction) throws StorageException
         try {
             originalAutoCommit = getDatabase().getAutoCommit();
             setDatabaseAutoCommit(false);
-
+            if (transaction.getId() == -1) {
+                transaction = getTransactionWithoutId(transaction);
+            }
+//            if(transaction.getType().equals(TypeConversion.ACC_TRANSFER)){
+//                if(!deleteOtherTransferTransaction(transaction))
+//                    throw new StorageException("Failed to create Other transfer transaction");
```
Comment (shellajt) [inline src/main/java/ledger/database/storage/SQL/ISQLDatabaseTransaction.java]: Lets remove these

File: src/main/java/ledger/database/storage/SQL/ISQLDatabaseTransaction.java [hunk 3]
```diff
@@ -84,7 +109,13 @@ default void deleteTransaction(Transaction transaction) throws StorageException
...
+//            }
```
Comment (shellajt) [inline src/main/java/ledger/database/storage/SQL/ISQLDatabaseTransaction.java]: Remove

File: src/main/java/ledger/user_interface/ui_controllers/component/UserTransactionInput.java [hunk 4]
```diff
@@ -40,6 +40,10 @@
     @FXML
     private TextField tagText;
     @FXML
+    private AccountDropdown account2Text;
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/component/UserTransactionInput.java]: Lets rename these to be meaningfull. sourceAccount and destinationAccount seem good

File: src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/CustomCells/TypeComboBoxTableCell.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import javafx.scene.text.Text;
+import javafx.util.StringConverter;
+import ledger.database.entity.Transaction;
+import ledger.database.entity.Type;
+import ledger.io.input.TypeConversion;
+
+
+/**
+ * Created by gert on 3/25/17.
+ */
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/CustomCells/TypeComboBoxTableCell.java]: Javadocs all up in here

File: src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/TypeColumn.java [hunk 6]
```diff
@@ -38,7 +42,13 @@ public void initialize(URL location, ResourceBundle resources) {
...
 
         ObservableList<Type> observableAllTypes = FXCollections.observableList(allTypes);
-        this.setCellFactory(ComboBoxTableCell.forTableColumn(new TypeStringConverter(), observableAllTypes));
+        this.setCellFactory( new Callback<TableColumn<Transaction,Type>, TableCell<Transaction,Type>>() {
+            @Override
+            public TableCell<Transaction,Type> call(TableColumn<Transaction,Type> param) {
+                return new TypeComboBoxTableCell(new TypeStringConverter(), observableAllTypes);
+            }
+        });
+        //.forTableColumn(new TypeStringConverter(), observableAllTypes));
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/TypeColumn.java]: Remove

File: src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/PayeeColumn.java [hunk 7]
```diff
@@ -25,26 +33,58 @@
...
+            }
+        });
         this.setOnEditCommit(new PayeeEventHandler());
         this.setComparator(new PayeeComparator());
 
         DbController.INSTANCE.registerPayeeSuccessEvent((ignored) -> Startup.INSTANCE.runLater(this::updatePayeeList));
+        DbController.INSTANCE.registerAccountSuccessEvent((ignored) -> Startup.INSTANCE.runLater(this::updateAccountList));
+    }
+
+    private void updateAccountList() {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/PayeeColumn.java]: Javadoc

File: src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/CustomCells/PayeeTableCell.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+import ledger.database.entity.IEntity;
+import ledger.database.entity.Payee;
+
+import java.util.ArrayList;
+
+/**
+ * Created by gert on 3/27/17.
+ */
+public class PayeeTableCell extends ComboBoxTableCell<Transaction,IEntity> {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/CustomCells/PayeeTableCell.java]: Javadoc

**Comment by shellajt** [comment]
I agree that we should notify users before making cascading changes.

I think we should disable a transfer from an account to to itself. You wouldn't want to do that on purpose

**Comment by taylerhow** [comment]
@shellajt Any thoughts on the cash flow I mentioned (with the photos)?

**Comment by shellajt** [comment]
@taylerhow it looks fine to me. Money is deposited into the account labeled "To"

**Comment by taylerhow** [comment]
@shellajt You are correct. I must've been looking at the wrong column (Payee instead of Account) earlier. My bad!

**Pull request merged**